packages/remirror__extension-file/src/file-loader.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * `FileLoader` is an interface based on the Web API `FileReader`. In remirror,
+ * the `FileReader` is used by default to read files uploaded by users, but if
+ * you don't want to use the browser's `FileReader`, you can design your own
+ * `FileReader`, as long as it implement this interface.
+ *
+ * Read more about the Web API `FileReader` here: https://developer.mozilla.org/en-US/docs/Web/API/FileReader
+ */
+export interface FileLoader {
+  /**
+   * A `DOMException` representing the error that occurred while reading the file.
+   */
+  readonly error: DOMException | null;
+
+  /**
+   * The file's contents. This property is only valid after the read operation
+   * is complete, and the format of the data depends on which of the methods was
+   * used to initiate the read operation.
+   */
+  readonly result: string | ArrayBuffer | null;
+
+  /**
+   * Aborts the read operation.
+   */
+  abort: () => void;
+
+  addEventListener: (
```

Comment:
Loving the adapter pattern here, however I think `addEventListener`, `loadend` etc is too closely coupled to `FileReader`'s implementation, perhaps we should be a bit more generic.

For instance [CKEditor has a nice adapter pattern](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html#the-anatomy-of-the-adapter) where the only things you need to implement are `upload` and `abort`. 

I'm not saying this is exactly the interface we should implement, just an example of a more generic design.

Reply:
Check the [comment](https://github.com/remirror/remirror/pull/1031#issuecomment-894934050) for a new API